### ISSUE TITLE
Inconsistencies between the confusion matrix of the training output and the output of the terminal

### ISSUE THREAD

**Issue by jiandaoni** [ultralytics/ultralytics#21007]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Comparing the two graphs, it can be seen that in the confusion matrix, the sensitivity (numerically equal to the recall) of the fire and smoke categories is 0.65 and 0.50, respectively, while the recall of the terminal output is 0.635 and 0.497, respectively. How can I make the output of both the same
![Image](https://github.com/user-attachments/assets/db85aec1-e0a0-4407-aba6-7fb945bf0cf7)

![Image](https://github.com/user-attachments/assets/ac82777f-1651-4ac4-abad-cba80359b318)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @jiandaoni, thank you for bringing this to our attention and for your interest in Ultralytics 🚀! This is an automated response to help guide you while an Ultralytics engineer reviews your issue and will assist you further soon.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) that demonstrates the inconsistencies you observed. This will help us investigate and debug the issue more efficiently.

For more information on usage and common questions, feel free to check out the [Docs](https://docs.ultralytics.com/), including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) guides.

Join the Ultralytics community wherever you prefer:
- For real-time chat, visit [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth discussions, check out [Discourse](https://community.ultralytics.com/)
- For community threads, join our [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

Please ensure you are running the latest `ultralytics` package with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) installed in a [**Python>=3.8**](https://www.python.org/) environment and [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO models can be run in the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
You can check this

https://github.com/ultralytics/ultralytics/issues/17047#issuecomment-2424660435

**Comment by jiandaoni** [comment]
Thank you for the recovery, in fact it is

**Comment by glenn-jocher** [comment]
Great that the linked explanation helped clarify this! Just to confirm - the slight differences you're seeing (0.65 vs 0.635, 0.50 vs 0.497) are expected because [confusion matrix precision focuses on overall classification accuracy across all classes and does not consider the geometric accuracy of predictions](https://docs.ultralytics.com/guides/yolo-common-issues/#understanding-precision-metrics-in-yolo11), while box metrics evaluate both classification and localization accuracy using IoU thresholds. These metrics assess different aspects of model performance, which is why they don't match exactly.

**Comment by kodaigit** [comment]
@glenn-jocher 
I am looking at this script. https://github.com/ultralytics/ultralytics/blob/6149368391f770dc41c720aa1644c052927bb2ff/ultralytics/utils/metrics.py#L348

To construct a confusion matrix, is the IoU threshold indeed used? (set to 0.45). So, I am guessing the confusion matrix considers the geometric accuracy as well. I might have missed something, though.

**Comment by jiandaoni** [comment]
I see this as well, it stands to reason that the confusion matrix should have some correlation with the box metrics, both are calculated from the predicted results of the dataset, what parameters should I set or what should I do if I want the confusion matrix to be as consistent as possible with the curves of these metrics.
https://github.com/ultralytics/ultralytics/issues/17047#issuecomment-2424660435
In this comment you just talk about calculations being different, can you be more specific?

**Comment by Y-T-G** [comment]
You can't make the confusion matrix the same as the metrics in logs because they're calculated differently. The confidence threshold is fixed in confusion matrix. While the logged metrics use a confidence threshold that's based on F1-score.

**Comment by jiandaoni** [comment]
Then I can use the confidence threshold based on the F1 score to output the confusion matrix. Under this condition, can the values ​​of P, R and other indicators calculated by the confusion matrix correspond to the values ​​on the curve?

**Comment by glenn-jocher** [comment]
@jiandaoni You're correct that both confusion matrix and evaluation metrics use IoU thresholds for matching predictions to ground truth. The key difference is in confidence thresholds: the confusion matrix uses a fixed confidence threshold (typically 0.25), while the logged metrics use optimal confidence thresholds that maximize F1-score for each class.

However, you cannot directly set the confusion matrix to use F1-score based confidence thresholds because the confusion matrix is generated during validation with a single fixed threshold, whereas the optimal thresholds are determined after analyzing the full precision-recall curve across all confidence levels. The curve metrics represent the best possible performance at optimal operating points, while the confusion matrix shows performance at your chosen fixed threshold.
